### PR TITLE
feat: Product Value Dataloader

### PR DESCRIPTION
**Pull request by aheykamp2u**
Adds the Product Value Dataloader

CSV header fields:
uuid - The UUID of the product
product type - 'course' or 'program'
per click usa - integer
per click international - integer
per lead usa - integer
per lead international - integer

Includes a configuration file model.

### CODE REVIEW

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+class ProductValueCSVDataLoader(AbstractDataLoader):
+    """Loads product value data from a csv file"""
+
+    PRODUCT_VALUE_REQUIRED_DATA_FIELDS = [
+        'uuid', 'product_type',
+    ]
+
+    PRODUCT_VALUE_OPTIONAL_DATA_FIELDS = [
+        'per_click_usa', 'per_click_international', 'per_lead_usa', 'per_lead_international'
+    ]
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: Well, if atleast one field is required, the word OPTIONAL misleads.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        """
+        if not self.is_valid_uuid(data['uuid']):
+            # invalid uuid captured in another validator
+            return
+
+        model = None
+        if data['product_type'] == 'course':
+            model = Course
+        elif data['product_type'] == 'program':
+            model = Program
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: nit: use dict to create this mapping and remove these if-elif conditions.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        logger.info("Product Value CSV loader ingest pipeline has completed.")
+
+        logger.info("Checking for orphaned product value records.")
+
+        for product_value in self.updated_product_values:
+            related_course = Course.everything.filter(in_year_value=product_value).first()
+            related_program = Program.objects.filter(in_year_value=product_value).first()
+            if (related_course is None and related_program is None):
+                ProductValue.objects.filter(id=product_value.id).delete()
+                logger.info("Removed orphaned product value with id: %i", product_value.id)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: nit: I would suggest moving this to a method to better explain why is this necessary.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        self.skipped_items = []
+        self.processed_courses = []
+        self.processed_programs = []
+        self.updated_product_values = []
+        try:
+            self.reader = csv.DictReader(open(csv_path, 'r')) if csv_path \
+                else list(unicodecsv.DictReader(csv_file))  # lint-amnesty, pylint: disable=consider-using-with
+
+        except FileNotFoundError:
+            logger.exception("Error opening csv file at path %s", csv_path)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: Python has f-strings in format`f"Error opening csv file at path {csv_path}"`. I will suggest updating all logs statements in this file to use f-strings. That is the new & consistent standard.

File: course_discovery/apps/course_metadata/management/commands/import_product_value_data.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        try:
+            loader = ProductValueCSVDataLoader(partner, csv_path=csv_path, csv_file=csv_file)
+            logger.info("Starting CSV loader import")
+            loader.ingest()
+        except Exception as exc:
+            raise CommandError(  # pylint: disable=raise-missing-from
+                "Product Value CSV loader import could not be completed due to unexpected errors.\n{}".format(exc)
+            )
+        logger.info("CSV loader import flow completed.")
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/management/commands/import_product_value_data.py]: nit:  ProductValue CSV loader flow

File: course_discovery/apps/course_metadata/data_loaders/tests/test_product_value_csv_loader.py [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+            csv = self._write_csv(
+                csv, [valid_product_value_csv_dict]
+            )
+            with LogCapture(self.LOGGER_PATH) as log_capture:
+                loader = ProductValueCSVDataLoader(self.partner, csv_path=csv.name)
+                loader.ingest()
+                if product_type == 'program':
+                    product = Program.objects.get(uuid=valid_uuid)
+                else:
+                    product = Course.everything.get(draft=False, uuid=valid_uuid)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/tests/test_product_value_csv_loader.py]: nit: if draft=False, then use Course.objects.get(). In this case, draft is filtered out automatically and there won't be a need for this if condition. The product_model can be defined as ddt param

**Review by DawoudSheraz — COMMENTED** [review]
Left a few small comments.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py (hunk 2 above)
Comment (aheykamp2u) [reply course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: Not exactly sure how to do what you're asking here.  Can you provide an example?

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
+"""
+Data loader responsible for creating product value entries in discovery database,
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: nit: redundant comma `,`

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        if not self.is_valid_uuid(data['uuid']):
+            # invalid uuid captured in another validator
+            return
+
+        models = {'course': Course, 'program': Program}
+        try:
+            model = models[data['product_type']]
+        except KeyError:
+            # invalid product_type error_msg captured in another validator
+            return
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: We can add a generic error message in `error_msgs`

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        if not is_course_or_program:
+            error_msgs.append(f"Wrong product_type value for UUID: {data['uuid']}")
+
+    def validate_product_exists(self, data, error_msgs):
+        """
+        Get an object from the database by its key and value
+        """
+        if not self.is_valid_uuid(data['uuid']):
+            # invalid uuid captured in another validator
+            return
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: here as well.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+            model = models[data['product_type']]
+        except KeyError:
+            # invalid product_type error_msg captured in another validator
+            return
+
+        model_name = model._meta.object_name
+        kwrags = {'uuid': data['uuid']}
+        try:
+            if model_name == 'Course':
+                exists = model.everything.filter(**kwrags).exists()
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: We can directly use this in the next condition. `exists` seems like an unnecessary variable.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+                if not exists:
+                    error_msgs.append(f"{data['product_type'].capitalize()} with UUID: {data['uuid']} was not found")
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: Since this error message `f"{data['product_type'].capitalize()} with UUID: {data['uuid']} was not found"` is seen 3 times, we can set this in a variable.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+
+    def process_program(self, data):
+        program_obj = Program.objects.filter(uuid=data['uuid']).first()
+        if program_obj.in_year_value:
+            self.updated_product_values.append(program_obj.in_year_value)
+            product_value_data = self.generate_values_dict(data, program_obj.in_year_value)
+            program_obj.in_year_value = ProductValue.objects.create(**product_value_data)
+            program_obj.save()
+            self.log_info(
+                f"Updated product value data for program with UUID: {data['uuid']}",
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: we can do the same `{action_taken}` thing here as well the same way you did that above.

File: course_discovery/apps/course_metadata/management/commands/import_product_value_data.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+        try:
+            partner = Partner.objects.get(short_code='edx')
+        except Partner.DoesNotExist:
+            raise CommandError(  # pylint: disable=raise-missing-from
+                "Unable to locate partner with code {}".format('edx')
+            )
+
+        try:
+            loader = ProductValueCSVDataLoader(partner, csv_path=csv_path, csv_file=csv_file)
+            logger.info("Starting CSV loader import")
```
Comment (Ali-D-Akbar) [inline course_discovery/apps/course_metadata/management/commands/import_product_value_data.py]: Product Value CSV loader import*

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py (hunk 8 above)
Comment (aheykamp2u) [reply course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: This error message is already caught further up the chain by another check. Adding it here might be redundant.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+        is_valid_uuid = self.is_valid_uuid(data['uuid'])
+        if not is_valid_uuid:
+            error_msgs.append(f"Invalid UUID: {data['uuid']}")
+
+    def validate_at_least_1_optional_field(self, data, error_msgs):
+        has_at_least_one_optional_field = False
+
+        for field in self.PRODUCT_VALUE_DATA_FIELDS:
+            if (field in data and data[field]):
+                has_at_least_one_optional_field = True
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: nit: use break to stop this loop as we have found one item.

File: course_discovery/apps/course_metadata/data_loaders/product_value_loader.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+            if model_name == 'Course':
+                exists = model.everything.filter(**kwrags).exists()
+                if not exists:
+                    error_msgs.append(f"{data['product_type'].capitalize()} with UUID: {data['uuid']} was not found")
+            else:
+                exists = model.objects.filter(**kwrags).exists()
+                if not exists:
+                    error_msgs.append(f"{data['product_type'].capitalize()} with UUID: {data['uuid']} was not found")
+        except model.DoesNotExist:
+            error_msgs.append(f"{data['product_type'].capitalize()} with UUID: {data['uuid']} was not found")
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/data_loaders/product_value_loader.py]: This is not necessary because we are using filter() and not get()

**Review by DawoudSheraz — APPROVED** [review]
Mostly, good. Once review/nit comments are addressed, this will be merged.

**Review by AliAdnanSohail — APPROVED** [review]
